File: sdk/javascript/src/command-execution.ts
/**
 * Class representing command execution results in a sandbox environment.
 */

import { OutputData, OutputLine } from "./types";

export class CommandExecution {
  private outputLines: OutputLine[] = [];
  private _command: string = "";
  private _args: string[] = [];
  private _exitCode: number = -1;
  private _success: boolean = false;

  /**
   * Initialize a command execution instance.
   *
   * @param outputData - Output data from the sandbox.command.run response
   */
  constructor(outputData?: OutputData) {
    if (outputData) {
      this.processOutputData(outputData);
    }
  }

  /**
   * Process output data from the sandbox.command.run response.
   *
   * @param outputData - Dictionary containing the output data
   */
  private processOutputData(outputData: OutputData): void {
    // Extract output lines from the response
    this.outputLines = outputData.output || [];

    // Store command-specific metadata
    this._command = outputData.command || "";
    this._args = outputData.args || [];
    this._exitCode =
      outputData.exit_code !== undefined ? outputData.exit_code : -1;
    this._success = outputData.success || false;
  }

  /**
   * Get the standard output from the command execution.
   *
   * @returns String containing the stdout output of the command
   */
  async output(): Promise<string> {
    // Combine the stdout output lines into a single string
    let outputText = "";
    for (const line of this.outputLines) {
      if (line.stream === "stdout") {
        outputText += line.text + "\n";
      }
    }

    return outputText.trim();
  }

  /**
   * Get the error output from the command execution.
   *
   * @returns String containing the stderr output of the command
   */
  async error(): Promise<string> {
    // Combine the stderr output lines into a single string
    let errorText = "";
    for (const line of this.outputLines) {
      if (line.stream === "stderr") {
        errorText += line.text + "\n";
      }
    }

    return errorText.trim();
  }

  /**
   * Get the exit code of the command execution.
   *
   * @returns Integer containing the exit code
   */
  get exitCode(): number {
    return this._exitCode;
  }

  /**
   * Check if the command executed successfully.
   *
   * @returns Boolean indicating whether the command succeeded (exit code 0)
   */
  get success(): boolean {
    return this._success;
  }

  /**
   * Get the command that was executed.
   *
   * @returns String containing the command
   */
  get command(): string {
    return this._command;
  }

  /**
   * Get the arguments used for the command.
   *
   * @returns List of strings containing the command arguments
   */
  get args(): string[] {
    return this._args;
  }
}

File: sdk/go/command_execution.go
package msb

import (
	"encoding/json"
	"strings"
)

// CommandExecution represents the result of command execution in the sandbox.
// Use the Get* methods for parsed access to output, or access Output directly for raw JSON.
type CommandExecution struct {
	Output    json.RawMessage // Raw JSON response from the server
	parsed    commandData     // Parsed data for convenience methods
	parsedOK  bool           // Whether parsing succeeded
}

// Internal structure for parsing command execution results
type commandData struct {
	OutputLines []outputLine `json:"output"`
	Command     string       `json:"command"`
	Args        []string     `json:"args"`
	ExitCode    int          `json:"exit_code"`
	Success     bool         `json:"success"`
}

// GetOutput returns the standard output from command execution as a string.
// Returns ErrExecutionNotParsed if the raw JSON could not be parsed.
func (ce CommandExecution) GetOutput() (string, error) {
	if !ce.parsedOK {
		return "", ErrExecutionNotParsed
	}
	
	var output strings.Builder
	for _, line := range ce.parsed.OutputLines {
		if line.Stream == "stdout" {
			output.WriteString(line.Text)
			output.WriteString("\n")
		}
	}
	return strings.TrimSuffix(output.String(), "\n"), nil
}

// GetError returns the error output from command execution as a string.
// Returns ErrExecutionNotParsed if the raw JSON could not be parsed.
func (ce CommandExecution) GetError() (string, error) {
	if !ce.parsedOK {
		return "", ErrExecutionNotParsed
	}
	
	var errorOutput strings.Builder
	for _, line := range ce.parsed.OutputLines {
		if line.Stream == "stderr" {
			errorOutput.WriteString(line.Text)
			errorOutput.WriteString("\n")
		}
	}
	return strings.TrimSuffix(errorOutput.String(), "\n"), nil
}

// GetExitCode returns the exit code of the executed command.
// Returns -1 if the raw JSON could not be parsed.
func (ce CommandExecution) GetExitCode() int {
	if !ce.parsedOK {
		return -1
	}
	return ce.parsed.ExitCode
}

// IsSuccess reports whether the command executed successfully (exit code 0).
// Returns false if the raw JSON could not be parsed.
func (ce CommandExecution) IsSuccess() bool {
	if !ce.parsedOK {
		return false
	}
	return ce.parsed.Success
}

// GetCommand returns the command that was executed.
// Returns empty string if the raw JSON could not be parsed.
func (ce CommandExecution) GetCommand() string {
	if !ce.parsedOK {
		return ""
	}
	return ce.parsed.Command
}

// GetArgs returns the arguments used for the command.
// Returns nil if the raw JSON could not be parsed.
func (ce CommandExecution) GetArgs() []string {
	if !ce.parsedOK {
		return nil
	}
	return ce.parsed.Args
}